semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -94,12 +100,66 @@ type mapping = Lval_env.t D.mapping
 (* HACK: Tracks tainted functions intrafile. *)
 type fun_env = (var, Taints.t) Hashtbl.t
 
+(* NOTE "Top sinks":
+ * (See note "Taint-tracking via ranges" in 'Match_tainting_mode.ml' for context.)
+ *
+ * Since we use sub-range checks to determine wheter a piece of code is a sink this
+ * can lead to unintuitive results. For example, `sink(if tainted then ok1 else ok2)`
+ * is reported as a tainted sink despite `if tainted then ok1 else ok2` is actually
+ * not tainted. The problem is that `tainted` is inside what `sink(...)` matches,
+ * and using a sub-range check we end up considering `tainted` itself as a sink.
+ *
+ * Unfortunately we cannot simply check for exact matches, because sometimes we
+ * cannot match sinks exactly. So, what we do is to check the top-level nodes in the
+ * CFG for sinks, and from all those matches we keep the "best" (larger) fits. We
+ * call these the "top sinks" (matches). Then we use those top sinks as a practical
+ * definition of "exact match", thus if we find  another match which is a strict
+ * sub-range of one of these top sinks, we simply disregard it. In our example above
+ * the top sink match will be `sink(if tainted then ok1 else ok2)`, so we will
+ * disregard `tainted` as a sink because we know there is a better match.
+ *)
+module Top_sinks = struct
+  module S = Set.Make (struct
+    type t = R.taint_sink tmatch
+
+    let compare m1 m2 =
+      let sink_id_cmp = String.compare m1.spec.R.sink_id m2.spec.R.sink_id in
+      if sink_id_cmp <> 0 then sink_id_cmp
+      else
+        let r1 = m1.range in
+        let r2 = m2.range in
+        if Range.(r1 $<=$ r2 || r2 $<=$ r1) then 0 else Stdlib.compare r1 r2
+  end)
+
+  type t = S.t
+
+  let empty = S.empty
+
+  let add m' sinks =
+    match S.find_opt m' sinks with
+    | None -> S.add m' sinks
+    | Some m ->
+        let r = m.range in
+        let r' = m'.range in
+        if r'.start > r.start || r'.end_ < r.end_ then
+          (* The new match is a worse fit so we keep the current one. *)
+          sinks
+        else (* We found a better (larger) match! *)
```

Comment:
If the rule defines two sinks, and one match is for one sink and another match is for another sink, then you will never reach this case because the `sink_id`s will be different.